src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -4661,18 +4670,15 @@ MacroAssembler::KlassDecodeMode MacroAssembler::klass_decode_mode() {
     return (_klass_decode_mode = KlassDecodeZero);
   }
 
-  if (operand_valid_for_logical_immediate(
-        /*is32*/false, (uint64_t)CompressedKlassPointers::base())) {
-    const uint64_t range_mask =
-      (1ULL << log2i(CompressedKlassPointers::range())) - 1;
-    if (((uint64_t)CompressedKlassPointers::base() & range_mask) == 0) {
+  if (int bits = eor_compatible_klass_encoding((uint64_t)CompressedKlassPointers::base())) {
+    if (CompressedKlassPointers::range() <= (1ULL << bits)) {
       return (_klass_decode_mode = KlassDecodeXor);
     }
   }
 
   const uint64_t shifted_base =
     (uint64_t)CompressedKlassPointers::base() >> CompressedKlassPointers::shift();
-  guarantee((shifted_base & 0xffff0000ffffffff) == 0,
+  guarantee(movk_compatible_klass_encoding(shifted_base),
             "compressed class base bad alignment");
```

Comment:
I also think this code can benefit from a bit cleanup some time in the future, but for I ask you to hold off: there is no pressing need at the moment, and once Lilliput things have settled, this code should be easier to work with it. 

Also, there are a few more things we could improve, and I rather see those in a combined single RFE.